pkg/kubectl/cmd/label.go

```diff
@@ -152,52 +150,80 @@ func labelFunc(obj runtime.Object, overwrite bool, resourceVersion string, label
 }
 
 func RunLabel(f *Factory, out io.Writer, cmd *cobra.Command, args []string) error {
-	if len(args) < 2 {
-		return util.UsageError(cmd, "<resource> <name> is required")
+	resources, labelArgs := []string{}, []string{}
+	first := true
+	for _, s := range args {
+		isLabel := strings.Contains(s, "=") || strings.HasSuffix(s, "-")
+		switch {
+		case first && isLabel:
+			first = false
+			fallthrough
+		case !first && isLabel:
+			labelArgs = append(labelArgs, s)
+		case first && !isLabel:
+			resources = append(resources, s)
+		case !first && !isLabel:
+			return util.UsageError(cmd, "all resources must be specified before label changes: %s", s)
+		}
 	}
-	if len(args) < 3 {
+	if len(resources) < 1 {
+		return util.UsageError(cmd, "one or more resources must be specified as <resource> <name> or <resource>/<name>")
+	}
+	if len(labelArgs) < 1 {
 		return util.UsageError(cmd, "at least one label update is required.")
 	}
-	res := args[:2]
+
+	selector := util.GetFlagString(cmd, "selector")
+	all := util.GetFlagBool(cmd, "all")
+	overwrite := util.GetFlagBool(cmd, "overwrite")
+	resourceVersion := util.GetFlagString(cmd, "resource-version")
+
 	cmdNamespace, err := f.DefaultNamespace()
 	if err != nil {
 		return err
 	}
 
-	mapper, _ := f.Object()
-	// TODO: use resource.Builder instead
-	mapping, namespace, name, err := util.ResourceFromArgs(cmd, res, mapper, cmdNamespace)
-	if err != nil {
-		return err
-	}
-	client, err := f.RESTClient(mapping)
+	labels, remove, err := parseLabels(labelArgs)
 	if err != nil {
 		return err
 	}
 
-	labels, remove, err := parseLabels(args[2:])
-	if err != nil {
+	mapper, typer := f.Object()
+	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand(cmd)).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		SelectorParam(selector).
+		ResourceTypeOrNameArgs(all, resources...).
+		Flatten().
+		Latest()
+
+	one := false
+	r := b.Do().IntoSingular(&one)
+	if err := r.Err(); err != nil {
 		return err
 	}
-	overwrite := util.GetFlagBool(cmd, "overwrite")
-	resourceVersion := util.GetFlagString(cmd, "resource-version")
+	// only apply resource version locking on a single resource
+	if !one {
+		resourceVersion = ""
```

Comment:
I think it's clearer to exit with an error if the user specified both resource version and multiple resources. Silently ignoring the resource version might confuse.